File: CS/DXExample.Module.Web/WebCheckedListBoxPropertyEditor.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.ExpressApp.Web.Editors.ASPx;
using DevExpress.Xpo;
using System.Web.UI.WebControls;
using DevExpress.ExpressApp.Web;
using DevExpress.Web.ASPxEditors;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.ExpressApp.Model.NodeWrappers;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Model;

namespace DXExample.Module.Web {
    [PropertyEditor(typeof(XPBaseCollection), false)]
    public class WebCheckedListBoxPropertyEditor : ASPxPropertyEditor, IComplexPropertyEditor {
        public WebCheckedListBoxPropertyEditor(Type objectType, IModelMemberViewItem model) : base(objectType, model) { }
        private TableEx table;
        XafApplication application;
        protected override WebControl CreateEditModeControlCore() {
            return CreateTable();
        }
        protected override WebControl CreateViewModeControlCore() {
            return CreateTable();
        }
        private Table CreateTable() {
            table = RenderHelper.CreateTable();
            table.ID = "MainTable";
            table.CssClass = "GroupContent";
            return table;
        }
        protected override void SetupControl(WebControl control) {
            base.SetupControl(control);
            if (ViewEditMode == DevExpress.ExpressApp.Editors.ViewEditMode.View) {
                control.Enabled = false;
            }
        }
        LightDictionary<ASPxCheckBox, object> items;
        XPBaseCollection checkedItems;
        protected override void ReadEditModeValueCore() {
            base.ReadEditModeValueCore();
            GenerateContent();
        }
        protected override void ReadViewModeValueCore() {
            base.ReadViewModeValueCore();
            GenerateContent();
        }
        protected override void SetImmediatePostDataScript(string script) { }
        protected override void SetImmediatePostDataCompanionScript(string script) { }
        private void GenerateContent() {
            table.Rows.Clear();
            items = new LightDictionary<ASPxCheckBox, object>();
            if (PropertyValue is XPBaseCollection) {
                checkedItems = (XPBaseCollection)PropertyValue;
                XPCollection dataSource = new XPCollection(checkedItems.Session, MemberInfo.ListElementType);
                IMemberInfo defaultMember = null;
                ITypeInfo typeInfo = MemberInfo.ListElementTypeInfo;
                IModelClass classInfo = application.Model.BOModel.GetClass(typeInfo.Type);
                if (!String.IsNullOrEmpty(classInfo.DefaultProperty)) {
                    defaultMember = typeInfo.FindMember(classInfo.DefaultProperty);
                }
                if (checkedItems.Sorting.Count > 0) {
                    dataSource.Sorting = checkedItems.Sorting;
                } else if (!String.IsNullOrEmpty(classInfo.DefaultProperty)) {
                    dataSource.Sorting.Add(new SortProperty(classInfo.DefaultProperty, DevExpress.Xpo.DB.SortingDirection.Ascending));
                }
                foreach (object obj in dataSource) {
                    CreateRow(obj, defaultMember);
                }
                foreach (object obj in checkedItems) {
                    CheckItem(obj);
                }
            }
        }
        private void CreateRow(object obj, IMemberInfo defaultMember) {
            TableRow row = new TableRow();
            row.ID = GetId("Row", obj);
            table.Rows.Add(row);
            TableCell valueCell = new TableCell();
            valueCell.CssClass = "Caption";
            valueCell.Width = Unit.Percentage(95);
            Literal value = new Literal();
            string displayText = String.Empty;
            if (defaultMember != null && !typeof(PersistentBase).IsAssignableFrom(defaultMember.MemberType)) {
                displayText = Convert.ToString(defaultMember.GetValue(obj));
            } else {
                displayText = obj.ToString();
            }
            value.Text = displayText;
            valueCell.Controls.Add(value);
            TableCell checkBoxCell = new TableCell();
            checkBoxCell.ID = GetId("CheckBoxCell", obj);
            ASPxCheckBox checkBox = RenderHelper.CreateASPxCheckBox();
            checkBox.ID = GetId("CheckBox", obj);
            checkBox.CheckedChanged += checkBox_CheckedChanged;
            items.Add(checkBox, obj);
            checkBoxCell.Controls.Add(checkBox);
            row.Cells.Add(checkBoxCell);
            row.Cells.Add(valueCell);
        }
        private void CheckItem(object obj) {
            ASPxCheckBox checkBox = items.FindKeyByValue(obj);
            checkBox.CheckedChanged -= checkBox_CheckedChanged;
            checkBox.Checked = true;
            checkBox.CheckedChanged += checkBox_CheckedChanged;
        }

        void checkBox_CheckedChanged(object sender, EventArgs e) {
            ASPxCheckBox editor = (ASPxCheckBox)sender;
            object obj = items[editor];
            if (editor.Checked) {
                checkedItems.BaseAdd(obj);
            } else {
                checkedItems.BaseRemove(obj);
            }
            OnControlValueChanged();
        }
        private string GetId(string elementName, object obj) {
            return elementName + MemberInfo.ListElementTypeInfo.KeyMember.GetValue(obj).ToString();
        }
        #region IComplexPropertyEditor Members

        public void Setup(IObjectSpace objectSpace, XafApplication application) {
            this.application = application;
        }

        #endregion
    }
}

File: VB/DXExample.Module.Web/WebCheckedListBoxPropertyEditor.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.ExpressApp.Web.Editors.ASPx
Imports DevExpress.Xpo
Imports System.Web.UI.WebControls
Imports DevExpress.ExpressApp.Web
Imports DevExpress.Web.ASPxEditors
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.ExpressApp.Model.NodeWrappers
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Model

Namespace DXExample.Module.Web
	Public Class WebCheckedListBoxPropertyEditor
		Inherits ASPxPropertyEditor
		Implements IComplexPropertyEditor
		Public Sub New(ByVal objectType As Type, ByVal model As IModelMemberViewItem)
			MyBase.New(objectType, model)
		End Sub
		Private table As TableEx
		Private application As XafApplication
		Protected Overrides Function CreateEditModeControlCore() As WebControl
			Return CreateTable()
		End Function
		Protected Overrides Sub SetImmediatePostDataScript(ByVal script As String)
		End Sub
		Protected Overrides Sub SetImmediatePostDataCompanionScript(ByVal script As String)
		End Sub
		Protected Overrides Function CreateViewModeControlCore() As WebControl
			Return CreateTable()
		End Function
		Private Function CreateTable() As Table
			table = RenderHelper.CreateTable()
			table.ID = "MainTable"
			table.CssClass = "GroupContent"
			Return table
		End Function
		Protected Overrides Sub SetupControl(ByVal control As WebControl)
			MyBase.SetupControl(control)
			If ViewEditMode = DevExpress.ExpressApp.Editors.ViewEditMode.View Then
				control.Enabled = False
			End If
		End Sub
		Private items As LightDictionary(Of ASPxCheckBox, Object)
		Private checkedItems As XPBaseCollection
		Protected Overrides Sub ReadEditModeValueCore()
			MyBase.ReadEditModeValueCore()
			GenerateContent()
		End Sub
		Protected Overrides Sub ReadViewModeValueCore()
			MyBase.ReadViewModeValueCore()
			GenerateContent()
		End Sub
		Private Sub GenerateContent()
			table.Rows.Clear()
			items = New LightDictionary(Of ASPxCheckBox, Object)()
			If TypeOf PropertyValue Is XPBaseCollection Then
				checkedItems = CType(PropertyValue, XPBaseCollection)
				Dim dataSource As New XPCollection(checkedItems.Session, MemberInfo.ListElementType)
				Dim defaultMember As IMemberInfo = Nothing
				Dim typeInfo As ITypeInfo = MemberInfo.ListElementTypeInfo
				Dim classInfo As IModelClass = application.Model.BOModel.GetClass(typeInfo.Type)
				If (Not String.IsNullOrEmpty(classInfo.DefaultProperty)) Then
					defaultMember = typeInfo.FindMember(classInfo.DefaultProperty)
				End If
				If checkedItems.Sorting.Count > 0 Then
					dataSource.Sorting = checkedItems.Sorting
				ElseIf (Not String.IsNullOrEmpty(classInfo.DefaultProperty)) Then
					dataSource.Sorting.Add(New SortProperty(classInfo.DefaultProperty, DevExpress.Xpo.DB.SortingDirection.Ascending))
				End If
				For Each obj As Object In dataSource
					CreateRow(obj, defaultMember)
				Next obj
				For Each obj As Object In checkedItems
					CheckItem(obj)
				Next obj
			End If
		End Sub
		Private Sub CreateRow(ByVal obj As Object, ByVal defaultMember As IMemberInfo)
			Dim row As New TableRow()
			row.ID = GetId("Row", obj)
			table.Rows.Add(row)
			Dim valueCell As New TableCell()
			valueCell.CssClass = "Caption"
			valueCell.Width = Unit.Percentage(95)
			Dim value As New Literal()
			Dim displayText As String = String.Empty
			If defaultMember IsNot Nothing AndAlso (Not GetType(PersistentBase).IsAssignableFrom(defaultMember.MemberType)) Then
				displayText = Convert.ToString(defaultMember.GetValue(obj))
			Else
				displayText = obj.ToString()
			End If
			value.Text = displayText
			valueCell.Controls.Add(value)
			Dim checkBoxCell As New TableCell()
			checkBoxCell.ID = GetId("CheckBoxCell", obj)
			Dim checkBox As ASPxCheckBox = RenderHelper.CreateASPxCheckBox()
			checkBox.ID = GetId("CheckBox", obj)
			AddHandler checkBox.CheckedChanged, AddressOf checkBox_CheckedChanged
			items.Add(checkBox, obj)
			checkBoxCell.Controls.Add(checkBox)
			row.Cells.Add(checkBoxCell)
			row.Cells.Add(valueCell)
		End Sub
		Private Sub CheckItem(ByVal obj As Object)
			Dim checkBox As ASPxCheckBox = items.FindKeyByValue(obj)
			RemoveHandler checkBox.CheckedChanged, AddressOf checkBox_CheckedChanged
			checkBox.Checked = True
			AddHandler checkBox.CheckedChanged, AddressOf checkBox_CheckedChanged
		End Sub

		Private Sub checkBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
			Dim editor As ASPxCheckBox = CType(sender, ASPxCheckBox)
			Dim obj As Object = items(editor)
			If editor.Checked Then
				checkedItems.BaseAdd(obj)
			Else
				checkedItems.BaseRemove(obj)
			End If
			OnControlValueChanged()
		End Sub
		Private Function GetId(ByVal elementName As String, ByVal obj As Object) As String
			Return elementName & MemberInfo.ListElementTypeInfo.KeyMember.GetValue(obj).ToString()
		End Function
		#Region "IComplexPropertyEditor Members"

		Public Sub Setup(ByVal objectSpace As IObjectSpace, ByVal application As XafApplication) Implements IComplexPropertyEditor.Setup
			Me.application = application
		End Sub

		#End Region
	End Class
End Namespace
